pgrx/src/callconv.rs

```diff
@@ -283,3 +283,182 @@ where
+// Potentially <'srf, 'curr, 'ret: 'curr + 'srf> -> <'ret> but don't start with that.
+// at first <'curr> or <'fcx>
+// It's a heap-allocated stack frame for your function.
+// 'fcx would apply to the arguments into this whole thing.
+// PgHeapTuple is super not ideal and this enables a replacement of that.
+// ArgAbi and unbox from fcinfo index
+// Just this and the accessors, something that goes from raw_args(&'a self) -> &'fcx [NullableDatum]? &'a [NullableDatum]?
+impl<'fcx> FcInfo<'fcx> {
+    /// Constructor, used to wrap a raw FunctionCallInfo provided by Postgres.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe as we cannot ensure the `fcinfo` argument is a valid
+    /// [`pg_sys::FunctionCallInfo`] pointer.  This is your responsibility.
+    pub unsafe fn assume_valid(fcinfo: pg_sys::FunctionCallInfo) -> FcInfo<'fcx> {
+        let _nullptr_check =
+            std::ptr::NonNull::new(fcinfo).expect("fcinfo pointer must be non-null");
+        Self(fcinfo, std::marker::PhantomData)
+    }
+    /// Retrieve the arguments to this function call as a slice of [`pgrx_pg_sys::NullableDatum`]
+    #[inline]
+    pub fn raw_args<'a>(&'a self) -> &'fcx [pgrx_pg_sys::NullableDatum]
+    where
+        'a: 'fcx,
+    {
+        // Null pointer check already performed on immutable pointer
+        // at construction time.
+        unsafe {
+            let arg_len = (*self.0).nargs;
+            let args_ptr: *const pg_sys::NullableDatum = std::ptr::addr_of!((*self.0).args).cast();
+            // A valid FcInfoWrapper constructed from a valid FuntionCallInfo should always have
+            // at least nargs elements of NullableDatum.
+            std::slice::from_raw_parts(args_ptr, arg_len as _)
+        }
+    }
+
+    /// Modifies the contained `fcinfo` struct to flag its return value as null.
+    ///
+    /// # Examples
+    ///
+    /// ```rust,no_run
+    /// use pgrx::pg_return_null;
+    /// use pgrx::prelude::*;
+    /// fn foo(fcinfo: pg_sys::FunctionCallInfo) -> pg_sys::Datum {
+    ///     return unsafe { pg_return_null(fcinfo) };
+    /// }
+    /// ```
+    #[inline]
+    pub unsafe fn pg_return_null(&mut self) -> pg_sys::Datum {
+        let fcinfo = unsafe { self.0.as_mut() }.unwrap();
+        fcinfo.isnull = true;
+        pg_sys::Datum::from(0)
+    }
+
+    /// Get the collation the function call should use
+    #[inline]
+    pub unsafe fn pg_get_collation(&self) -> pg_sys::Oid {
+        let fcinfo = unsafe { self.0.as_mut() }.unwrap();
+        fcinfo.fncollation
+    }
+
+    /// Retrieve the type (as an Oid) of argument number `num`.
+    /// In other words, the type of `self.raw_args()[num]`
+    #[inline]
+    pub fn pg_getarg_type(&self, num: usize) -> pg_sys::Oid {
+        // Safety: User must supply a valid fcinfo to assume_valid() in order
+        // to construct a FcInfo. If that constraint is maintained, this should
+        // be safe.
```

Comment:
You can just say something like
```suggestion
        // SAFETY: see FcInfo::assume_valid
```